### PR TITLE
Added Services and Utilities Tests to Core.Test project

### PR DESCRIPTION
**Pull request by justindbaur**
Description
I cleaned up some of the tests I add in PR #1062, I sorted the usings, used an underscore on private fields, and used the system under test naming scheme. I also added a test class for the `CoreHelpers`, I did not test every single method in this class and some of the tests are not super thorough but it appears they are quite used so it might be useful to even have the minimum. 

Related #419

Notes
When testing the `CoreHelpers.ReadableBytesSize` method, I setup a theory that used some inline data that is, to put it lightly very out of bounds. Since the lone parameter is long I put through both `long.MinValue` and `long.MaxValue`.  Since `long.MaxValue` translated to over 9000 petabytes I have a feeling there won't be an offering of this size coming any time soon.  But that seemed to work just fine, `long.MinValue` on the other hand didn't seem to do what I thought it would do based on the code.  Line 330 of `CoreHelpers.cs` seems to want to absolute value the incoming value but it just leaves the negative value alone and from there goes through the logic of formatting it.  It returns the negative number followed by "Bytes".  I have left the few tests commented out if anyone wants to see them.  If any of this seems like an issue you want address please let me know and I can see what I can do.

### CODE REVIEW

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            Assert.Equal(item0, dt.Rows[0][0]);
+            Assert.Equal(item1, dt.Rows[1][0]);
+        }
+
+        // TODO: Test the other ToArrayTVP Methods
+
+        [Theory]
+        [InlineData("12345&6789", "123456789")]
+        [InlineData("abcdef", "ABCDEF")]
+        [InlineData("1!@#$%&*()_+", "1")]
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:64]: Please add a unicode character test to this.

I believe the original purpose of the method was to clean out the unicode lef-right-mark control character so something like `[InlineData("\u200E1234", "1234")]` should be good.

Comment (MGibson1) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:64]: While you're at it, why not add some other characters that could mess things up? I'm thinking
* `\n`
* `\t`
* '\r'
* etc

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            // Arrange & Act
+            var @string = CoreHelpers.SecureRandomString(8);
+
+            // Assert
+            // TODO: Should probably add more Asserts down the line
+            Assert.Equal(8, @string.Length);
+        }
+
+        [Theory]
+        // [InlineData(long.MinValue, "8589934592 GB")]
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:109]: Remove unused tests

Add string tests for each transition
* Kilobyte (`0x400`)
* Megabyte (`0x100000`)
* Gigabyte (`0x40000000`)

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        {
+            var orignial = new { Message = "Message" };
+
+            var copy = CoreHelpers.CloneObject(orignial);
+
+            Assert.Equal(orignial.Message, copy.Message);
+        }
+
+        [Fact]
+        public void ExtendQuery_Success()
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:132]: I think It's worth quickly extending this to the following cases to develop a bit more of a contract

* uri with no parameters
* overriding a param value
* empty dictionary doesn't change the uri

**Review by MGibson1 — CHANGES_REQUESTED** [review]
Thank you for this submission.

I have a few comments on the new tests to expand cases (if you're up for it! 😄) and on the Service test stubs.

The value of those stub test is also pretty low while using AutoFixture. It may make more sense to just remove them if there are no tests beyond creation of a SUT.

I'm more than happy to discuss and help with any of this, just let me know!

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 2 above)
Comment (MGibson1) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:109]: I forgot about your note on these commented tests. The size number is based on the `size` variable, not the `absoluteSize` variable. That's why the result was still negative.

I think that's *probably* intended, but it's not clear to me what a negative size string would be useful for

File: test/Core.Test/Services/AppleIapServiceTests.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                {
+                    AppInReview = true,
+                    Password = "AppleIapServicesTestsPassword",
+                }
+            };
+
+            _webHostEnvironment = Substitute.For<IWebHostEnvironment>();
+            _metaDateRepository = Substitute.For<IMetaDataRepository>();
+
+            _sut = new AppleIapService(_globalSettings,
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply test/Core.Test/Services/AppleIapServiceTests.cs:34]: @MGibson1 I have not had a chance to work in AutoFixture quite yet but would be happy to learn.  I will probably add a commit to remove these scaffolded tests, then work on adding them back one at a time as full AutoFixture implementations, those will probably need some reviews so I will look forward to some tips and tricks on those if you are up to it.  Thanks!

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 1 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:64]: I will have a commit to address this and will mark this as resolved when complete.

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 2 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:109]: I did not notice that the number is based on the size variable but the suffix to use is still based on the `absoluteSize`.  `long.MinValue` is still not flipped since is is one smaller than `long.MaxValue` is large.  Example: https://dotnetfiddle.net/NBB4kt

But since this is still a super edge case I will remove the unused tests and add in a few more useful ones, just wanted to bring it up since I noticed.  I will add some more tests and mark this as resolved.

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 3 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:132]: This is a very good idea, I will add all those and try to push it to it's limits a bit more.

File: test/Core.Test/Services/AppleIapServiceTests.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+using NSubstitute;
+using Xunit;
+
+namespace Bit.Core.Test.Services
+{
+    public class AppleIapServiceTests
+    {
+        private readonly AppleIapService _sut;
+
+        private readonly GlobalSettings _globalSettings;
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply test/Core.Test/Services/AppleIapServiceTests.cs]: I have removed this test for now, but may add it back in when I figure out AutoFixture, the interfaces conversion might still be a good one to have though.

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            // Arrange & Act
+            var comb = CoreHelpers.GenerateComb();
+
+            // Assert
+            Assert.NotEqual(Guid.Empty, comb);
+            // TODO: Add more asserts to make sure important aspects of
+            // the comb are working properly
+        }
+
+        // TODO: Probably make this a Theory with some more possibilties.
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs]: Added a commit to address this.

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+        [Theory]
+        [InlineData(1, "1 Bytes")]
+        [InlineData(-5L, "5 Bytes")]
+        [InlineData(1023L, "1023 Bytes")]
+        [InlineData(1024L, "1 KB")]
+        [InlineData(1025L, "1 KB")]
+        [InlineData(-1023L, "1023 Bytes")]
+        [InlineData(-1024L, "-1 KB")]
+        [InlineData(-1025L, "-1 KB")]
```
Comment (justindbaur) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:138]: Something to note here. For anything that is less than a KB the absolute size is used in the formatting so therefore even putting in -1023 gives a positive 1023 Bytes as opposed to passing in negative for any larger numbers.

**Review by justindbaur — COMMENTED** [review]
I think this is what I am supposed to do to say I have reviewed the changes you requested and that I believe I have address them.  (Not a pro on PR reviews in GitHub though, so if I shouldn't have done this whoops)

**Comment by MGibson1** [comment]
> I think this is what I am supposed to do to say I have reviewed the changes you requested and that I believe I have address them. (Not a pro on PR reviews in GitHub though, so if I shouldn't have done this whoops)

No, sorry, the request review you did is all that's needed. I just haven't had time to take a look yet today. I'm looking now though ;)

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 8]
```diff
@@ -20,20 +20,37 @@ public void GenerateComb_Success()
...
+        [InlineData(2, 5, new[] { 1, 2, 3, 4, 5, 6, 7, 8 , 9, 0 })]
+        [InlineData(2, 3, new[] { 1, 2, 3, 4, 5 })]
+        [InlineData(2, 1, new[] { 1, 2 })]
+        [InlineData(1, 1, new[] { 1 })]
+        [InlineData(2, 2, new[] { 1, 2, 3 })]
+        public void Batch_Success(int batchSize, int totalBatches, int[] collection)
         {
             // Arrange
-            var source = new[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+            var hasRemainder = collection.Length % batchSize != 0;
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:32]: Sorry to needle here, but this can be done with smaller branches.

```
var remainder = collection.Length % batchSize;

var batches = collection.Batch(batchSize);

Assert.Equal(totalBatches, batches.Count());

foreach(var batch in batches.Take(totalBatches - 1))
{
    Assert.Equal(batchSize, batch.Count());
}

Assert.Equal(batches.Last(), remainder == 0 ? batchSize : remainder)
```

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 9]
```diff
@@ -106,11 +128,26 @@ public void SecureRandomString_Success()
         }
 
         [Theory]
-        // [InlineData(long.MinValue, "8589934592 GB")]
-        // [InlineData(long.MinValue + 1, "Change")]
-        [InlineData(1000L, "1000 Bytes")]
         [InlineData(1, "1 Bytes")]
         [InlineData(-5L, "5 Bytes")]
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:132]: If you don't mind, we should make this method consistent by using the size variable for all the stringifications.

 I can commit the change if you don't want to deal with it.

**Review by MGibson1 — CHANGES_REQUESTED** [review]
This is looking really good. I had two more little things and then I think it's g2g.

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 8 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:32]: Perfect, I will make that change.

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 9 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs:132]: I agree, I would be more than happy to make that change.

File: test/Core.Test/Utilities/CoreHelpersTests.cs [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+            // Act & Assert
+            Assert.Equal(milliseconds, CoreHelpers.ToEpocMilliseconds(date));
+        }
+
+        [Theory]
+        [InlineData(1609310952000L, "2020-12-30T11:49:12.0000000Z")]
+        public void FromEpocMilliseconds(long milliseconds, string dateString)
+        {
+            // Arrange
+            var date = DateTime.Parse(dateString);
```
Comment (MGibson1) [inline test/Core.Test/Utilities/CoreHelpersTests.cs:105]: I'm sorry about this but I'd assumed out pipelines were running on this PR and they aren't.

This doesn't pass for me locally. It seems DateTime.Parse is creating a local kind datetime despite the 'Z' in your string. ugh, I hate dates AND times.

The solution that worked for me locally was to use the [MemberData] provider and instantiate the DateTime as `new DateTime(2020,12,30,11,49,12, DateTimeKind.Utc)`. Then I think we can be sure it works regardless of local.

```
public static IEnumerable<object[]> _epochTestCases = new[] {
    new object[] {new DateTime(2020, 12, 30, 11, 49, 12, DateTimeKind.Utc), 1609328952000L},
};

// TODO: Add more tests
[Theory]
[MemberData(nameof(_epochTestCases))]
public void ToEpocMilliseconds_Success(DateTime date, long epochMilliseconds)
{
    // Act & Assert
    Assert.Equal(epochMilliseconds, CoreHelpers.ToEpocMilliseconds(date));
}

[Theory]
[MemberData(nameof(_epochTestCases))]
public void FromEpocMilliseconds(DateTime date, long epochMilliseconds)
{
    // Act & Assert
    Assert.Equal(date, CoreHelpers.FromEpocMilliseconds(epochMilliseconds));
}
```

**Review by MGibson1 — CHANGES_REQUESTED** [review]
I'm sorry I should have noticed this before, but there are failing tests due to localizations

File: test/Core.Test/Utilities/CoreHelpersTests.cs (hunk 10 above)
Comment (justindbaur) [reply test/Core.Test/Utilities/CoreHelpersTests.cs]: Out of curiosity I was wondering what it would take to get a Utc date from a string and this is what I came up with.  I still like your solution much better so I have implemented it but figured I would show you. 

```
var date = DateTime.ParseExact("2020-12-30T11:49:12.0000Z", 
    "yyyy-MM-ddThh:mm:ss.ffffZ", 
    CultureInfo.InvariantCulture, 
    DateTimeStyles.AdjustToUniversal);
```

**Review by MGibson1 — APPROVED** [review]
Thank you for your contribution! These tests are very nice and excellent additions to shore up some foundational code 👍

**Review by kspearrin — APPROVED** [review]

**Comment by justindbaur** [comment]
@MGibson1 Thanks! I will work on some service tests utilizing AutoFixture, I may open up a draft PR if I have any questions on how to implement them to Bitwarden standards.  Have a great day!
